### PR TITLE
Move combinations, permutations, etc., from Array to Indexable

### LINKED ISSUE #10063: [RFC] Move combinations, permutations, etc., from Array to Indexable
Today I was working with a `Deque` and I wanted to use `each_combination(2)` with it... but it doesn't work! It's only defined in Array.

I think this is because these methods predate the introduction of `Indexable`.

It would be great if we move these methods to `Indexable`. Then you could ask all permutations over a Tuple, Deque, Slice, StaticArray, and more!

Of course the question is: what does these methods return? In the case of `Array` they return `Array`. Should they return a `Deque` in the case of `Deque`? Or a `Tuple` in the case of a `Tuple`? Well, the answer to the second question is easy: there's no way to return `Tuple` because the return value's size is "kind of" dynamic: it depends on the number passed to the method, which is a runtime value.

With that in mind, I think they should always return `Array`: `Array` is the default "list-like" collection type.

So, to implement this we would just need to move these methods from `Array` to `Indexable`.

### PR DESCRIPTION
**Pull request by wonderix**
I tried to implement #10063 to help judging if it's feasible.

Moving `Array.product` to `Indexable` made no sense to me.  Therefore it was also not possible to move `Array#repeated_permutations` to `Indexable`

Fixes #10063

### CODE REVIEW

File: src/indexable.cr [hunk 1]
```diff
@@ -618,6 +618,258 @@ module Indexable(T)
...
 
+  # Returns *n* number of random elements from `self`, using the given *random* number generator.
+  # Raises IndexError if `self` is empty.
+  #
+  # ```
+  # a = [1, 2, 3]
+  # a.sample(2)                # => [2, 1]
+  # a.sample(2, Random.new(1)) # => [1, 3]
+  # ```
+  def sample(n : Int, random = Random::DEFAULT)
```
Comment (HertzDevil) [inline src/indexable.cr:629]: `#sample` is already covered by #10129.

**Comment by HertzDevil** [comment]
`#repeated_permutations` can be implemented on top of #10013.

File: src/indexable.cr (hunk 1 above)
Comment (wonderix) [reply src/indexable.cr]: I reverted the changes for `sample`

File: src/indexable.cr [hunk 2]
```diff
@@ -668,4 +882,172 @@ module Indexable(T)
...
+    size.times { |i| reuse << pool[i] }
+    reuse
+  else
+    pool[0, size]
+  end
+end
+
+private def dup_as_array(a)
+  pool = a.to_a
+  pool.object_id == a.object_id ? pool.dup : pool
```
Comment (HertzDevil) [inline src/indexable.cr:1052]:
```suggestion
  a.is_a?(Array) ? a.dup : a.to_a
```

`#object_id` is not defined on most structs like `Tuple` and `StaticArray`.

**Review by straight-shoota — APPROVED** [review]

**Review by asterite — APPROVED** [review]

**Comment by asterite** [comment]
@wonderix This is great, thank you!

**Pull request merged**